core/app/models/spree/order.rb

```diff
@@ -788,10 +788,9 @@ def ensure_inventory_units
     end
 
     def ensure_promotions_eligible
-      adjustment_changed = all_adjustments.eligible.promotion.any? do |adjustment|
-        !adjustment.calculate_eligibility
-      end
-      if adjustment_changed
+      Spree::Config.promotion_adjuster_class.new(self).call
```

Comment:
What [this class](https://github.com/solidusio/solidus/blob/00031a2c1844e881a08128847289924573bb6b48/core/app/models/spree/promotion/order_adjustments_recalculator.rb#L17-L41) does seems pretty different from the `calculate_eligibility` method's content. Can you help me understand in which way they are equivalent in this context?

Reply:
It does check the eligibility of each adjustment in `recalculate(adjustment)` and especially [`calculate_eligibility(adjustment)`](https://github.com/solidusio/solidus/blob/00031a2c1844e881a08128847289924573bb6b48/core/app/models/spree/promotion/order_adjustments_recalculator.rb#L83-L89). The class does a little bit more - it calculates not only eligibility, but also checks whether the adjustment's sum of amount is still correct. So it actually checks the promo total that the error message indicates it should have checked.

Reply:
I got it; I am a bit worried about degrading the performance of this method with the call to the `promotion_adjuster_class` because I guess (I didn't check) it will hit the database many more times. Considering that it only happens on the order completion, it might be a good trade-off. If you or others are not concerned about that, I'm okay with merging this change.

Reply:
There's a slight performance degradation for sure. From my experience with the promotion system though, the expensive part is checking the rules for eligibility, and not (re)calculating the discounts. 
Also: I fully agree that it's a tradeoff, and that it's okay because it only happens on order completion.